File: Linked List/2. Doubly Linked List/2. Doubly LinkedList Insertion/DoublyLinkedListInsertion.cpp
#include <iostream>
using namespace std;

class DLLNode
{
    public:
        int data;
        DLLNode* nextElement;
        DLLNode* prevElement;

        DLLNode()
        {
            nextElement = nullptr;
            prevElement = nullptr;
        }
};

class DoublyLinkedList
{
    private:
        DLLNode* head;
    
    public:
        DoublyLinkedList()
        {
            head = nullptr;
        }

        bool isEmpty()
        {
            if (head == nullptr)
                return true;
            else
            {
                return false;
            }
            
        }

        void printDLL()
        {
            if (isEmpty())
            {
                cout << "Doubly DoublyLinkedList is empty." << endl;
            }

            DLLNode* temp = head;
            while (temp != nullptr)
            {
                cout << temp->data << "->";
                temp = temp->nextElement;
            }
            cout << "null " << endl;
        }

        void insertAtHead(int value)
        {
            DLLNode* newNode = new DLLNode();
            newNode->data = value;

            newNode->nextElement = head;

            if (head != nullptr)
                head->prevElement = newNode;
            
            head = newNode;
            cout << value << " inserted." << endl;

        }

        void insertAtTail(int value)
        {
            DLLNode* newNode = new DLLNode();
            newNode->data = value;

            DLLNode* temp = head;

            while (temp->nextElement != nullptr)
                temp = temp->nextElement;

            temp->nextElement = newNode;
            newNode->prevElement = temp;

            cout << value << " inserted." << endl;
        }
};

int main()
{
    DoublyLinkedList list; // DoublyLinkedList created

    for (int i = 0; i < 5; i++)
    {
        list.insertAtHead(i+1);
        list.printDLL();
    }

    for (int i = 30; i > 25; i--)
    {
        list.insertAtTail(i);
        list.printDLL();
    }

    return 0;
}

File: Linked List/2. Doubly Linked List/2. Doubly LinkedList Insertion/DoublyLinkedListInsertion.java
class DLLNode
{
    int value;
    DLLNode next;
    DLLNode prev;

    DLLNode()
    {
        this.next = null;
        this.prev = null;
    }

    DLLNode(int value)
    {
        this.value = value;
        this.next = null;
        this.prev = null;
    }
}

public class DoublyLinkedListInsertion
{
    private DLLNode head;

    public DoublyLinkedListInsertion()
    {
        head = null;
    }

    public boolean isEmpty()
    {
        if (head == null)
            return true;
        else
            return false;
    }

    public void printList()
    {
        if (isEmpty())
            System.out.println("LinkedList is Empty");
        
        DLLNode temp = head;

        while (temp != null)
        {
            System.out.print(temp.value + "->");
            temp = temp.next;
        }
        System.out.println("null");
    }

    public void insertAtHead(int value)
    {
        DLLNode newNode = new DLLNode(value);

        newNode.next = head;

        if (head != null)
            head.prev = newNode;
        
        head = newNode;
        System.out.println(value + " inserted." );
    }

    public void insertAtTail(int value)
    {
        DLLNode newNode = new DLLNode(value);
        
        if (isEmpty())
        {
            head = newNode;
            System.out.println(value + " inserted.");
            return;
        }

        DLLNode temp = head;
        while (temp.next != null)
        {
            temp = temp.next;
        }

        temp.next = newNode;
        newNode.prev = temp;

        System.out.println(value + " inserted.");
    }

    public static void main(String args[])
    {
        DoublyLinkedListInsertion list = new DoublyLinkedListInsertion();

        for (int i = 0; i < 10; i++)
        {
            list.insertAtHead(i);
            list.printList();
        }

        for (int i = 30; i > 20; i--)
        {
            list.insertAtTail(i);
            list.printList();
        }
    }
}